File: app/src/main/java/com/jerson/hcdc_portal/ui/adapter/BuildingAdapter.java
package com.jerson.hcdc_portal.ui.adapter;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.Transition;
import com.davemorrissey.labs.subscaleview.ImageSource;
import com.jerson.hcdc_portal.R;
import com.jerson.hcdc_portal.databinding.ItemContainerBuildingBinding;
import com.jerson.hcdc_portal.listener.OnClickListener;
import com.jerson.hcdc_portal.model.BuildingModel;

import java.util.List;

public class BuildingAdapter extends RecyclerView.Adapter<BuildingAdapter.ViewHolder> {

    private List<BuildingModel> list;
    private OnClickListener<Integer> listener;

    public BuildingAdapter(List<BuildingModel> list, OnClickListener<Integer> listener) {
        this.list = list;
        this.listener = listener;
    }

    @NonNull
    @Override

    public BuildingAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(ItemContainerBuildingBinding.inflate(LayoutInflater.from(parent.getContext()),parent,false));
    }

    @Override
    public void onBindViewHolder(@NonNull BuildingAdapter.ViewHolder holder, int position) {
        BuildingModel item = list.get(position);
        holder.binding.buildingName.setText(item.getName());

        Glide.with(holder.binding.buildingImage)
                .asBitmap()
                .load(item.getImage_prev())
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .error(Glide.with(holder.binding.getRoot())
                        .asBitmap()
                        .load(R.drawable.logo)
                        .diskCacheStrategy(DiskCacheStrategy.ALL)
                        .error(R.drawable.logo))
                .into(new CustomTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(@NonNull Bitmap resource, @Nullable Transition<? super Bitmap> transition) {
                        if(!resource.isRecycled()){
                            holder.binding.buildingImage.setImageBitmap(resource);
                        }else resource.recycle();
                    }

                    @Override
                    public void onLoadCleared(@Nullable Drawable placeholder) {

                    }
                });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final ItemContainerBuildingBinding binding;

        public ViewHolder(@NonNull ItemContainerBuildingBinding itemView) {
            super(itemView.getRoot());
            this.binding = itemView;

            binding.getRoot().setOnClickListener(v->{
                listener.onItemClick(getAdapterPosition());
            });
        }
    }
}

File: app/src/main/java/com/jerson/hcdc_portal/presentation/building/adapter/BuildingAdapter.kt
package com.jerson.hcdc_portal.presentation.building.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import coil.ImageLoader
import coil.load
import com.jerson.hcdc_portal.databinding.ItemContainerBuildingBinding
import com.jerson.hcdc_portal.domain.model.Building
import com.jerson.hcdc_portal.util.Constants

class BuildingAdapter(private val list:List<Building>,private val itemCallback:(Building)->Unit):RecyclerView.Adapter<BuildingAdapter.ViewHolder>(){
    inner class ViewHolder(private val binding: ItemContainerBuildingBinding):RecyclerView.ViewHolder(binding.root){
        fun bind(item:Building){
            binding.buildingName.text = item.name
            val imgLoader = ImageLoader.Builder(binding.root.context).allowHardware(false).build()
            binding.buildingImage.load("${Constants.githubContent}${item.img_path}",imgLoader)
            binding.buildingImage.setOnClickListener{
                itemCallback(item)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            ItemContainerBuildingBinding.inflate(
                LayoutInflater.from(parent.context),
                parent,
                false
            )
        )
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = list[position]
        holder.bind(item)
    }
}